game-app/game-core/src/main/java/games/strategy/triplea/util/UnitSeparator.java

```diff
@@ -48,36 +48,80 @@ public static class SeparatorCategories {
+    final GameData gameData = t.getData();
+    return getComparatorUnitCategories(t, gameData, gameData.getHistory().getCurrentPlayer());
+  }
+
+  /** Returns <code>Comparator</code> for unit categories with current <code>GameData</code> */
+  public static Comparator<UnitCategory> getComparatorUnitCategories(final GameData gameData) {
+    return getComparatorUnitCategories(null, gameData, gameData.getHistory().getCurrentPlayer());
+  }
+
+  /** Returns <code>Comparator</code> for unit categories of a <code>Territory</code> */
+  private static Comparator<UnitCategory> getComparatorUnitCategories(
+      final Territory t, final GameData gameData, final GamePlayer currentPlayer) {
+    final List<UnitType> xmlUnitTypes =
+        new ArrayList<>(gameData.getUnitTypeList().getAllUnitTypes());
+    final List<GamePlayer> players = gameData.getPlayerList().getPlayers();
+    return getComparatorUnitCategories(t, currentPlayer, players, xmlUnitTypes);
+  }
+
+  /**
+   * Returns <code>Comparator</code> for unit categories to allow sorting into logical order to
+   * display based on: 1. Unit owner: territory owner, not at war with territory owner, player order
+   * in XML 2. Unit type: 0 movement, can't combat move, sea, air if sea territory, land, air if
+   * land territory 3. Within each of those groups sort the units by XML order in UnitList
+   */
+  private static Comparator<UnitCategory> getComparatorUnitCategories(
+      final Territory t,
+      final GamePlayer currentPlayer,
+      final List<GamePlayer> players,
+      final List<UnitType> xmlUnitTypes) {
+    return Comparator.comparing(
+            UnitCategory::getOwner,
+            Comparator.comparing((final GamePlayer p) -> !(t != null && p.equals(t.getOwner())))
+                .thenComparing(p -> (t != null && Matches.isAtWar(p).test(t.getOwner())))
+                .thenComparing(players::indexOf))
+        .thenComparing(uc -> Matches.unitTypeCanMove(uc.getOwner()).test(uc.getType()))
+        .thenComparing(
+            UnitCategory::getType,
+            Comparator.comparing(
+                    (final UnitType ut) -> !Matches.unitTypeCanNotMoveDuringCombatMove().test(ut))
+                .thenComparing(ut -> !Matches.unitTypeIsSea().test(ut))
+                .thenComparing(
+                    ut -> !(t != null && t.isWater() && Matches.unitTypeIsAir().test(ut)))
+                .thenComparing(ut -> !Matches.unitTypeIsLand().test(ut)))
+        .thenComparingInt(ut -> ut.getUnitAttachment().getMaxBuiltPerPlayer())
+        .thenComparing(
+            uc ->
+                uc.getUnitAttachment()
+                    .getAttack(
+                        (currentPlayer == null
+                            ? uc.getOwner()
+                            : currentPlayer))) // should be currentPlayer
+        .thenComparing(UnitCategory::getType, Comparator.comparing(xmlUnitTypes::indexOf));
```

Comment:
The comparator logic is a bit dense. Some inline comments might help clean it up a bit. Fully spelled out variables might help as well (potentially usually logical names, like "attackingUnits". Not sure if those suggestions are going to move the needle. Overall this is a block to double check and re-read in case you see a good way to help clean it up a bit more.

Reply:
Wasn't too happy with it either, but then I thought as long as the method JavaDoc is explaining the logic I could understand the code as well.
I could offer to create separate Comparator instantiations that are then used in the return, but for me it wouldn't be really much better readable.

Reply:
Mentioned in another comment - if we can remove territory and player as inputs - this could be a bit simpler perhaps (?)
I think the key would be to tie the blocks together to the javadoc, or to simplify the javadoc and put more details inline. Naming the predicates can be of help too, eg:

```
Predicate<UnitCategory> hasMovement = uc -> ....
```

Reply:
I am not sure how to rework this. Shall we have a call instead of post-ping-pong here? Send me a pm with some time slots you can offer.

Reply:
I like the idea of shifting to a call. Though, I'm not sure of a good way to really fix this either! 

FWIW, I intended this to be the operative part of the review comment:
> Overall this is a block to double check and re-read in case you see a good way to help clean it up a bit more.

We can and should move on now that we've done that double check